numpyro/hmc_util.py

```diff
@@ -332,9 +330,8 @@ def _biased_transition_kernel(current_tree, new_tree):
     return transition_prob
 
 
-@partial(jit, static_argnums=(6,))
-def _combine_tree(current_tree, new_tree, inverse_mass_matrix, going_right, rng,
-                  biased_transition, iterative_build):
+@partial(jit, static_argnums=(5,))
+def _combine_tree(current_tree, new_tree, inverse_mass_matrix, going_right, rng, biased_transition):
```

Comment:
We can make `biased_transition` a dynamic arg instead of static and using various `lax.cond` inside the function. I have tried it but don't see any improvement in performance. Explicitly splitting this function into 2 versions (one for biased, one for uniform) also does not add any improvement. And it turns out at the end of this work that these versions just have a small overhead over HMC.

Reply:
Since compiled functions are cached, this function will get compiled at most twice, so I wouldn't necessarily expect that working around that would lead to any savings. Its great that you dug into this, and verified it though.